mirrord-layer/src/tcp/outgoing.rs

```diff
@@ -0,0 +1,308 @@
+                            break;
+                        },
+                        Ok(read_amount) => {
+                            // Sends the message that the user wrote to our interceptor socket to
+                            // be handled on the `agent`, where it'll be forwarded to the remote.
+                            let write = Write { connection_id, bytes: buffer[..read_amount].to_vec() };
+                            let outgoing_data = TcpOutgoing::Write(write);
+
+                            if let Err(fail) = send_hook_message(HookMessage::TcpOutgoing(outgoing_data)).await {
+                                error!("Failed sending write message with {:#?}!", fail);
+                                break;
+                            }
+                        }
+                    }
+                },
+                bytes = remote_stream.next() => {
+                    match bytes {
+                        Some(bytes) => {
+                            // Writes the data sent by `agent` (that came from the actual remote
+                            // stream) to our interceptor socket. When the user tries to read the
+                            // remote data, this'll be what they receive.
+                            if let Err(fail) = mirror_stream.write_all(&bytes).await {
+                                error!("Failed writing to mirror_stream with {:#?}!", fail);
+                                break;
+                            }
+                        },
+                        None => {
+                            warn!("interceptor_task -> exiting due to remote stream closed!");
+                            break;
+                        }
+                    }
+                },
+            }
+        }
+
+        trace!(
+            "interceptor_task done -> connection_id {:#?}",
+            connection_id
+        );
+    }
+
+    /// Handles the following hook messages:
+    ///
+    /// - `TcpOutgoing::Connect`: inserts the new connection request into a connection queue, and
+    ///   sends it to (agent) as a `TcpOutgoingRequest::Connect` with the remote host's address.
+    ///
+    /// - `TcpOutgoing::Write`: sends a `TcpOutgoingRequest::Write` message to (agent) with the data
+    ///   that our interceptor socket intercepted.
+    pub(crate) async fn handle_hook_message(
+        &mut self,
+        message: TcpOutgoing,
+        codec: &mut actix_codec::Framed<
+            impl tokio::io::AsyncRead + tokio::io::AsyncWrite + Unpin + Send,
+            ClientCodec,
+        >,
+    ) -> Result<(), LayerError> {
+        trace!("handle_hook_message -> message {:?}", message);
+
+        match message {
+            TcpOutgoing::Connect(Connect {
+                remote_address,
+                channel_tx,
+            }) => {
+                trace!("Connect -> remote_address {:#?}", remote_address);
+
+                self.connect_queue.push_back(channel_tx);
```

Comment:
Starting to get afraid by the way we do this queues instead of some sort of HashMap (there's an odd case for a race condition probably right now). Nothing that should be done right now, just a thought on the first thing we should refactor - some generic API for passing Request/Response between layer/agent that is ID based for avoiding evil

Reply:
`TODO` added.